# Hands-On Esxercise 1 - Business Anlaytics (Sales Performance Calculator)

Using NumPy to perform basic numerical analysis on the sales data of a growing retail business . The management team wants to understand the week's peformance.

# Task 1

In [5]:
# Import NumPY

import numpy as np

# create the daily_sales array
daily_sales = np.array([125000, 150000, 175000,140000, 190000, 210000, 160000])

# print the array
print("Daily Sales:", daily_sales)

# Print the type of the array
print("Daily Sales Type:", daily_sales.dtype)

# Shape of the array

print("Daily Sales Shape:", daily_sales.shape)

Daily Sales: [125000 150000 175000 140000 190000 210000 160000]
Daily Sales Type: int64
Daily Sales Shape: (7,)


# Task 2

In [6]:
# Calculate the total sales. To get total sales, the sum of the daily sales is needed

total_sales = np.sum(daily_sales)

# Print result
print("Total Sales:", total_sales)


Total Sales: 1150000


# Task 3

In [14]:
# Calculate Average Daily Sales

average_sales = np.sum(daily_sales)/ len(daily_sales)

# Print result
print("Average Sales:", average_sales)

Average Sales: 164285.7142857143


# Task 4

In [11]:
# Applying a sales adjustment. 

daily_sales = np.array([125000, 150000, 175000,140000, 190000, 210000, 160000])

# Rule: Each sales get 10% increase of the original price i.e original price + 10% increase
# A 10% increase = 100 + (10/100) = 1.10 
increase = 1.10

# create a new array to represent the adjusted sales by multiply each day sales by 1.10 

adjusted_sales = np.multiply(daily_sales, increase)

# Print result

print("Adjusted Sales:", adjusted_sales)

Adjusted Sales: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]


# Task 5

In [12]:
# Compare Original sales and the adjusted sales by calculting the additional revenue from 10% sales increase

daily_sales = np.array([125000, 150000, 175000,140000, 190000, 210000, 160000])

sales_difference = np.subtract(adjusted_sales, daily_sales)

# print result

print("Sales Difference:", sales_difference)

Sales Difference: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]


# Task 6

In [13]:
# Using NumPy Mathematical Functions to calculate the average sales.

average_sales = np.average(daily_sales)

# Print results
print("Total Sales:", total_sales)
print("Average Sales:", average_sales)

Total Sales: 1150000
Average Sales: 164285.7142857143


Explanation: The total sales show the combined revenue generated over the seven-day period, while the average dail sales shows the typical amount generated per day. The average makes it easier for the usiness to understand its daily sales performance an compare it with the performance of the other periods.

# Hands- On Exercise 2 - Education Analytics (Student Performance Analysis)

Using NumPy to perform numerical analysis for a training program that recorded the scores of a group of students . The program coordinator wants to understand the overall performance of the students and how individual scores differ from the group average.

# Task 1

In [16]:
# import NumPy as np

import numpy as np

# create the numpy array array

scores = np.array([62, 75,81,69,88,94,73,85,77,91])

# Print result
print("Student Scores:", scores)
print("Score Type:", scores.dtype)
print("Shape:", scores.shape)

Student Scores: [62 75 81 69 88 94 73 85 77 91]
Score Type: int64
Shape: (10,)


# Task 2

In [17]:
# Calculate the mean score

mean_score = np.mean(scores)

# Print result
print("Mean Score:", mean_score)

Mean Score: 79.5


# Task 3

In [18]:
# Calculate each student deviation.How far is each student's score from the average(mean)

scores = np.array([62, 75,81,69,88,94,73,85,77,91])
mean_score = np.mean(scores)

deviations = np.subtract(scores, mean_score)

# Print result

print("Deviations:", deviations)

Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


# Task 4

Observations from the Deviations:

Positive deviation, negative derivation  and deviation close to zero.

This means a positive deviation means that a student's score is above the mean, while a negative deviation means that the score is below the group mean. A deviation close to zero means that te student's score is close to the average score of the group.


# Task 5

In [19]:
# Create a new aray contatining the squared deviations.

scores = np.array([62, 75,81,69,88,94,73,85,77,91])
mean_score = np.mean(scores)

deviations = np.subtract(scores, mean_score)

# Squaring the deviations
squared_deviations = deviations ** 2

# Print result
print("Squared Deviations:", squared_deviations)

Squared Deviations: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]


The deviations are squared so that all the values become positive. This prevents positive and negative deviations from cancelling each other out and allows us to measure the size of the differences from the mean.

# Task 6

In [20]:
# Calculate the sum of squared deviations

squared_deviations = deviations ** 2

sum_squared_deviations = np.sum(squared_deviations)

# Print result

print("Sum of Squared Deviations:", sum_squared_deviations)


Sum of Squared Deviations: 932.5


# Task 7

## Connecting the steps to Standard Deviation

The standard deviation calculation involves several steps:

1. Calculate the mean of the scores.

2. Calculate the deviation of each score from the mean

3. Square each deviation.

4. Add the squared deviations together.

5. Divide the sum by the appropriate number of observations

6. Take the square root of the result

Standard deviation is useful because it shows how spread out the student scores are around the mean. While the mean score tells us the average performance, standard deviation gives additional information about how consistently the students performed. A smaller standard deviation indicates that the scores  are closer together, while a larger standard deviation indicates greater variation is performance.

Mean score = 79.5 ---> Deviations -----> squared deviations ----> sum of squared deviations = 932.5

## Industry Thinking Challenge

The average score alone is not enough because it ony shows the overall performance and does not show spread out thescores are. Standard deviation helps to show the variation in students' scores.

# Hands-On 3- Engineering & Scientific Computing (Trigonometric Series and Convergence

Using NumPy to create a sequence of values, work wirh a trigonometric function, and observe observe the idea of convergence. The team wants to understand how an approximation changes as more terms are included in a calculaion.


# Task 1

In [3]:
#import numpy as np

import numpy as np

# Define the angle

theta = 30

# convert the angle to radians

theta_radians = np.radians(30)

# print result
print("Angle:", theta_radians)

Angle: 0.5235987755982988


# Task 2

In [10]:
# Create a sequence of values for k. 

k = np.arange (1, 100, 1)

# print the resulting array for the sequence. Fist value, last value and the number of values generated

print("K Sequence:",k)
print("First Value:", k[0])
print("Last value:", k[-1])
print("Number of values:", len(k))

K Sequence: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]
First Value: 1
Last value: 99
Number of values: 99


# Task 3

In [11]:
#calculate the trigonometric component

sin_theta = np.sin(theta_radians)

# print result

print("Sin Theta:", sin_theta)

Sin Theta: 0.49999999999999994


# Task 4

In [13]:
# Building the series terms

sin_theta = np.sin(theta_radians)
k = np.arange (1, 100, 1)

# create the individual terms of the series by diving the sine component by k

terms = np.divide(sin_theta, k)

#print result

print("Terms:", terms)

Terms: [0.5        0.25       0.16666667 0.125      0.1        0.08333333
 0.07142857 0.0625     0.05555556 0.05       0.04545455 0.04166667
 0.03846154 0.03571429 0.03333333 0.03125    0.02941176 0.02777778
 0.02631579 0.025      0.02380952 0.02272727 0.02173913 0.02083333
 0.02       0.01923077 0.01851852 0.01785714 0.01724138 0.01666667
 0.01612903 0.015625   0.01515152 0.01470588 0.01428571 0.01388889
 0.01351351 0.01315789 0.01282051 0.0125     0.01219512 0.01190476
 0.01162791 0.01136364 0.01111111 0.01086957 0.0106383  0.01041667
 0.01020408 0.01       0.00980392 0.00961538 0.00943396 0.00925926
 0.00909091 0.00892857 0.00877193 0.00862069 0.00847458 0.00833333
 0.00819672 0.00806452 0.00793651 0.0078125  0.00769231 0.00757576
 0.00746269 0.00735294 0.00724638 0.00714286 0.00704225 0.00694444
 0.00684932 0.00675676 0.00666667 0.00657895 0.00649351 0.00641026
 0.00632911 0.00625    0.00617284 0.00609756 0.0060241  0.00595238
 0.00588235 0.00581395 0.00574713 0.00568182 0.00561798

# Task 5

In [14]:
# Calculate the sum

series_sum = np.sum(terms)

#print result

print(series_sum)

2.5886887588198104


# Task 6

In [15]:
#Increase the number of terms

# For 99 terms
k_99 = np.arange(1, 100, 1)
terms_99 = np.divide(sin_theta, k_99)
sum_99 = np.sum(terms_99)

# For 999 terms
k_999 = np.arange(1, 1000, 1)
terms_999 = np.divide(sin_theta, k_999)
sum_999 = np.sum(terms_999)

# For 9999 terms
k_9999 = np.arange(1, 10000, 1)
terms_9999 = np.divide(sin_theta, k_9999)
sum_9999 = np.sum(terms_9999)

# print result
print("99 terms:", sum_99)
print("999 terms:", sum_999)
print("9999 terms:", sum_9999)

99 terms: 2.5886887588198104
999 terms: 3.7422354302751724
9999 terms: 4.89375301802219


# Task 7

# Investigate convergence

Term (99) --------->>>> 2.5886887588198104

Term(999) ---------->>>>> 3.7422354302751724

Term(9999) --------->>>>>> 4.89375301802219

As the number of terms increases, the series result also increases. The additional terms contribute smaller values to the total, so the result changes progressively as more terms are included


# Task 8

# Convergence Explanation wth the experiment

Convergence means that as more terms are added, the result  approaches a limiting value. The additional terms become smaller, so their effect on the total gradually decreases. When the result gets closer to a limiting value, t indicates that the series is approaching convergence. 

Basically, when thr fewer terms are used in the series, the rsult is based on a smaller number of vales and gives a lower approximation. As the number of terms increases, more values are included in the calculation, causing the series result to change and ecome a better approximation of the series.

# BONUS CHALLENGE

# MINI NUMPY NUERICAL EXPERIMENT: ENERGY CONSUMPTION ANALYSIS

Scenario:

An energy analyst wants to examine the electricity consumption of a small facility over seven days.

Objective:

The objective is to use NumPy to calculate the total and average electricity consumption, estimate the variation in daily consumption, and examine how consumption changes across the seven-day period.


In [23]:
print("="* 50)
print()
print("    ENERGY CONSUMPTION ANALYSIS")
print("="*50)
print(f""" Scenario:

An energy analyst wants to examine the electricity consumption of a small facility over seven days.

Objective:

The objective is to use NumPy to calculate the total and average electricity consumption, estimate the variation in daily consumption, 
and examine how consumption changes across the seven-day period.
""")

print()

# Data Set + Code

# import numpy as np

import numpy as np

# Daily electricity consumption in kWh

consumption = np.array([120, 150, 135, 180, 160,200,175])

# Print result

print("Daily Consumption:", consumption)
print()
# calculate the total and average

total_consumption = np.sum(consumption)
average_consumption = np.average(consumption)

# Print result

print("Total Consumption:", total_consumption, "kWh")
print("Average Daily Consumption:", average_consumption, "kWh")
print()
#calculate for the devation

deviation = np.subtract(consumption, average_consumption)

# Print result

print("Deviation from Average:", deviation)
print()
# using np.sqrt()

squared_deviation = deviation** 2
mean_squared_deviation = np. mean(squared_deviation)
standard_deviation = np.sqrt(mean_squared_deviation)

# print result
print("Standard deviation:", standard_deviation, "kWh")
print()


print(F""" RESULTS AND INTERPRETATION
    Results
    
The total electricity consumption over the seven-day period was 1,120kWh, while the average daily consumption was 160kWh. 
The standard deviation was approximaely 25.49kWh.

    Interpretation

The facility consumed an average of 160kWh per day. Some days recorded consumption above the average , while others were below it. 
The standard deviation of approximately 25.49 kWh shows that daily electricity consumption w=varied around the average rather than remaining constant.

""")

print()
print(F""" CONCLUSION

This experiment demonstrates how NumPy can be used to analyse real-word energy consumpton data.
NumPy functions such as np.array(),np.sum(), and np.sqrt() made it possible to calculate and interpret important measures of electricity consumption 
efficiently.
""")

print("="*50)



    ENERGY CONSUMPTION ANALYSIS
 Scenario:

An energy analyst wants to examine the electricity consumption of a small facility over seven days.

Objective:

The objective is to use NumPy to calculate the total and average electricity consumption, estimate the variation in daily consumption, 
and examine how consumption changes across the seven-day period.


Daily Consumption: [120 150 135 180 160 200 175]

Total Consumption: 1120 kWh
Average Daily Consumption: 160.0 kWh

Deviation from Average: [-40. -10. -25.  20.   0.  40.  15.]

Standard deviation: 25.495097567963924 kWh

 RESULTS AND INTERPRETATION
    Results

The total electricity consumption over the seven-day period was 1,120kWh, while the average daily consumption was 160kWh. 
The standard deviation was approximaely 25.49kWh.

    Interpretation

The facility consumed an average of 160kWh per day. Some days recorded consumption above the average , while others were below it. 
The standard deviation of approximately 25.49 kWh 